qiskit_nature/transformers/active_space_transformer.py

```diff
@@ -0,0 +1,193 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The Active-Space Reduction interface."""
+
+from typing import List, Optional
+import logging
+import numpy as np
+
+from qiskit_nature import QMolecule, QiskitNatureError
+from .base_transformer import BaseTransformer
+
+logger = logging.getLogger(__name__)
+
+
+class ActiveSpaceTransformer(BaseTransformer):
+    """The Active-Space reduction."""
+
+    def __init__(self, num_electrons: int, num_orbitals: int, num_alpha: Optional[int] = None,
```

Comment:
I think whatever this acts on should maintain the num electrons etc and orbital indexes currently present in the system that may already have been transformed. Here in my mind I should just be specifying orbitals to be removed - whether they be occupied or unoccupied. If any orbitals do not exist at transform time then raise an error.

I was more thinking we would transform the quantities that were interested in an leave the raw driver result untouched. Otherwise reducing its AOs etc etc seems like unneccessry work. So if you consider this extra data in this class that contains the driver result as what is being transformed, then yes the raw result would be untransformed but this can be stated.

Reply:
It does not make sense to specify the active space based on the orbitals you would like to remove. As an example you can think of a large metal complex which has hundreds of orbitals. If I want to select an active space (of say 10 orbitals) I want to specify those and not the hundreds of inactive ones.

The most simplistic choice of active space is `CAS(#nelec, #norb)` which stands for "complete active space" with `#nelec` electrons in `#norbs` active orbitals. More sophisticated choices can be done by manually selecting orbital indices (which still need to match the number of electrons and orbitals, though) and for those we need to figure out how to keep track of possible index shifting due to other transforms. However, that part of the code can only be added once we have more than one working transformer.

As per AOs: I do have a TODO there but I don't think it really makes sense to transform them at all. Although if a user specifies and active space we may want to be able to specify in which basis this is supposed to happen. But we can add that in the future.

Reply:
Sure, I have no problem in having different mechanisms to specify it. I guess whether you do it on what you want to keep active or throw away depends on what is the larger quantity for active space, and exactly how you want to specify it whether things are contiguous spaces or need to be stated separately, And it would be nice whatever is done here is consistent with UCCSD.

I don't think it makes sense to transform AOs either - which is why I was proposing to keep the aspects that would be transformed more separate. That way the raw result stays as the raw driver output and is self-consistent in the set of data it contains with the other integrals we need being maintained elsewhere.

Reply:
The biggest difference between this transform and the old orbital reduction method is that here we go through the "effort" of correcting for the exclusion of the reduced orbitals. This means, that we incorporate an effective potential describing the excluded electrons into the active-space Hamiltonian and therefore don't introduce an approximation which may not be valid in certain edge cases.
Since this is a mathematically more correct approach which does not pose too much of a computational overhead, I would argue that this can replace the original orbital reduction fully.

However, if we find that it does indeed introduce an overhead which we do not want, we can always add a more simplistic orbital reduction transformer using the logic we had previously. However, then we will have to make sure to clearly differentiate the two and make sure that the "active space" notation is not mixed up and is consistent across the two transformers.

Reply:
Sounds fine to me if this is an improvement over what existed that retains the same level of reduction in problem complexity over what we finally compute.

I guess part of my initial comment was around the configuration where before we just specified orbitals that were wanted to be discarded (we did not have any option to state those to keep.) As such it was just a list of orbitals so you could pick stuff out wherever you wanted and it was frozen or deleted depending on whether said orbital was occupied or not.

One thing that would help, at least for me, is to improve the docstring to better describe the parameters and any interaction. In my mind all we have before is akin to the active_orbitals parameter that is defaulted to None (we specified the inactive ones to be discarded so those that were not to be kept as active). I guess since we know what is occupied or not from that list the num_electrons and num_orbitals seems confusing to me. Maybe its a simpler alternative way or specifying things but they are mandatory  parameters unlike the orbital list. Overall I was trying to figure what these do, as such, maybe the num_alpha is in there to designate which of the orbitals hold only alpha electrons out of the total number provided in the first param. Hence I had pointed out that all we did before was have a list that we deleted (maybe the deleted part took your thoughts in a different direction) . But I meant we just had a list of orbitals and could figure things based on known occupancy. Here I imagine you have more flexibility to define the active space and maybe all it needs for me is improved docstrings for the parameters - maybe we an inline example or two of configurations.

Reply:
Another thought on the orbital reduction that we currently have - is the extra info that is needed for the active space transform available in all case from the set of drivers. If not it might be good reason to have the simpler version too even if it is not as accurate?

Reply:
I have improved the class-level docstring to hopefully clarify the method a bit more. Let's discuss in a comment of a new review if we need to improve it further.

I will comment separately below about the freeze core/orbital reduciton method.